File: invertedai/large/_quadtree.py
from pydantic import BaseModel

import invertedai as iai
from invertedai.large.common import Region
from invertedai.common import Point, AgentState, AgentAttributes, RecurrentState

BUFFER_FOV = 35
QUADTREE_SIZE_BUFFER = 1

class QuadTreeAgentInfo(BaseModel):
    """
    All information relevant to a single agent.

    See Also
    --------
    AgentState
    AgentAttributes
    RecurrentState
    """

    agent_state: AgentState
    agent_attributes: AgentAttributes
    recurrent_state: RecurrentState
    agent_id: int

    def tolist(self):
        return [self.agent_state, self.agent_attributes, self.recurrent_state, self.agent_id]

    @classmethod
    def fromlist(cls, l):
        agent_state, agent_attributes, recurrent_state, agent_id = l
        return cls(agent_state=agent_state, agent_attributes=agent_attributes, recurrent_state=recurrent_state, agent_id=agent_id)

class QuadTree:
    def __init__(
        self, 
        capacity: int, 
        region: Region, 
    ):
        self.capacity = capacity
        self.region = region
        self.leaf = True
        self.northWest = None
        self.northEast = None
        self.southWest = None
        self.southEast = None

        self.region_buffer = Region.create_square_region(
            center=self.region.center,
            size=self.region.size+2*BUFFER_FOV
        )
        self.particles = []
        self.particles_buffer = []

    def subdivide(self):
        parent = self.region
        new_size = self.region.size/2
        new_center_dist = new_size/2
        parent_x = parent.center.x
        parent_y = parent.center.y

        region_nw = Region.create_square_region(
            center=Point.fromlist([parent_x-new_center_dist,parent_y+new_center_dist]),
            size=new_size
        )
        region_ne = Region.create_square_region(
            center=Point.fromlist([parent_x+new_center_dist,parent_y+new_center_dist]),
            size=new_size
        )
        region_sw = Region.create_square_region(
            center=Point.fromlist([parent_x-new_center_dist,parent_y-new_center_dist]),
            size=new_size
        )
        region_se = Region.create_square_region(
            center=Point.fromlist([parent_x+new_center_dist,parent_y-new_center_dist]),
            size=new_size
        )

        self.northWest = QuadTree(self.capacity,region_nw)
        self.northEast = QuadTree(self.capacity,region_ne)
        self.southWest = QuadTree(self.capacity,region_sw)
        self.southEast = QuadTree(self.capacity,region_se)

        self.leaf = False
        self.region.clear_agents()
        self.region_buffer.clear_agents()

        for particle in self.particles:
            is_inserted = self.insert_particle_in_leaf_nodes(particle,False)
        for particle in self.particles_buffer:
            is_inserted = self.insert_particle_in_leaf_nodes(particle,True)
        self.particles = []
        self.particles_buffer = []
        
    def insert_particle_in_leaf_nodes(self,particle,is_inserted):
        is_inserted_in_this_branch = self.northWest.insert(particle,is_inserted)
        is_inserted_in_this_branch = self.northEast.insert(particle,is_inserted_in_this_branch or is_inserted) or is_inserted_in_this_branch
        is_inserted_in_this_branch = self.southWest.insert(particle,is_inserted_in_this_branch or is_inserted) or is_inserted_in_this_branch
        is_inserted_in_this_branch = self.southEast.insert(particle,is_inserted_in_this_branch or is_inserted) or is_inserted_in_this_branch

        return is_inserted_in_this_branch

    def insert(self, particle, is_particle_placed=False):
        is_in_region = self.region.is_inside(particle.agent_state.center)
        is_in_buffer = self.region_buffer.is_inside(particle.agent_state.center)

        if (not is_in_region) and (not is_in_buffer):
            return False

        if (len(self.particles) + len(self.particles_buffer)) < self.capacity and self.leaf:
            if is_in_region and not is_particle_placed:
                self.particles.append(particle)
                self.region.insert_all_agent_details(*particle.tolist()[:-1])
                return True

            else: # Particle is within the buffer region of this leaf node
                self.particles_buffer.append(particle)
                self.region_buffer.insert_all_agent_details(*particle.tolist()[:-1])
                return False

        else:
            if self.leaf:
                self.subdivide()

            is_inserted = self.insert_particle_in_leaf_nodes(particle,is_particle_placed)

            return is_inserted

    def get_regions(self):
        if self.leaf:
            return [self.region]
        else:
            return self.northWest.get_regions() + self.northEast.get_regions() + \
                self.southWest.get_regions() + self.southEast.get_regions()

    def get_leaf_nodes(self):
        if self.leaf:
            return [self]
        else:
            return self.northWest.get_leaf_nodes() + self.northEast.get_leaf_nodes() + \
                self.southWest.get_leaf_nodes() + self.southEast.get_leaf_nodes()

    def get_number_of_agents_in_node(self):
        return len(self.particles)

def _flatten_and_sort(nested_list,index_list):
    flat_list = [x for sublist in nested_list for x in sublist]
    sorted_list = [x[1] for x in sorted(zip(index_list, flat_list))]
    
    return sorted_list

File: invertedai_cpp/large/quadtree.cc
#include "quadtree.h"
#include <cmath>
#include <algorithm>
#include <iostream>
#include <vector>

namespace invertedai {

QuadTree::QuadTree(int capacity, const Region& region)
    : capacity_(capacity),
      region_(region),
      region_buffer_(Region::create_square_region(region.center,
                                                  region.size + 2 * BUFFER_FOV)),
      leaf_(true),
      northWest_(nullptr),
      northEast_(nullptr),
      southWest_(nullptr),
      southEast_(nullptr) {}

void QuadTree::subdivide() {
    Region parent = region_;
    double new_size = region_.size / 2.0;
    double new_center_dist = new_size / 2.0;
    double parent_x = parent.center.x;
    double parent_y = parent.center.y;

    Region region_nw = Region::create_square_region({parent_x - new_center_dist,
                                                     parent_y + new_center_dist},
                                                    new_size);
    Region region_ne = Region::create_square_region({parent_x + new_center_dist,
                                                     parent_y + new_center_dist},
                                                    new_size);
    Region region_sw = Region::create_square_region({parent_x - new_center_dist,
                                                     parent_y - new_center_dist},
                                                    new_size);
    Region region_se = Region::create_square_region({parent_x + new_center_dist,
                                                     parent_y - new_center_dist},
                                                    new_size);

    northWest_ = std::make_unique<QuadTree>(capacity_, region_nw);
    northEast_ = std::make_unique<QuadTree>(capacity_, region_ne);
    southWest_ = std::make_unique<QuadTree>(capacity_, region_sw);
    southEast_ = std::make_unique<QuadTree>(capacity_, region_se);

    leaf_ = false;
    region_.clear_agents();
    region_buffer_.clear_agents();

    for (const auto& p : particles_)
        insert_particle_in_leaf_nodes(p, false);
    for (const auto& p : particles_buffer_)
        insert_particle_in_leaf_nodes(p, true);

    particles_.clear();
    particles_buffer_.clear();
}

bool QuadTree::insert_particle_in_leaf_nodes(
    const QuadTreeAgentInfo& particle, 
    bool is_inserted
) {
    bool inserted = northWest_->insert(particle, is_inserted);
    inserted = northEast_->insert(particle, inserted || is_inserted) || inserted;
    inserted = southWest_->insert(particle, inserted || is_inserted) || inserted;
    inserted = southEast_->insert(particle, inserted || is_inserted) || inserted;
    return inserted;
}

bool QuadTree::insert(
    const QuadTreeAgentInfo& particle, 
    bool is_particle_placed
) {
    bool is_in_region = region_.is_inside({particle.agent_state.x, particle.agent_state.y});
    bool is_in_buffer = region_buffer_.is_inside({particle.agent_state.x, particle.agent_state.y});

    if (!is_in_region && !is_in_buffer)
        return false;

    if ((particles_.size() + particles_buffer_.size()) < static_cast<size_t>(capacity_) && leaf_) {
        if (is_in_region && !is_particle_placed) {
            particles_.push_back(particle);
            region_.insert_all_agent_details(particle.agent_state,
                                             particle.agent_properties,
                                             particle.recurrent_state.value_or(std::vector<double>{}));
            return true;
        } else {
            particles_buffer_.push_back(particle);
            region_buffer_.insert_all_agent_details(particle.agent_state,
                                                    particle.agent_properties,
                                                    particle.recurrent_state.value_or(std::vector<double>{}));
            return false;
        }
    }

    if (leaf_) {
        subdivide();
    }

    bool inserted = insert_particle_in_leaf_nodes(particle, is_particle_placed);
    return inserted;
}

std::vector<Region> QuadTree::get_regions() const {
    if (leaf_) return {region_};

    auto nw = northWest_->get_regions();
    auto ne = northEast_->get_regions();
    auto sw = southWest_->get_regions();
    auto se = southEast_->get_regions();

    nw.insert(nw.end(), ne.begin(), ne.end());
    nw.insert(nw.end(), sw.begin(), sw.end());
    nw.insert(nw.end(), se.begin(), se.end());
    return nw;
}

std::vector<QuadTree*> QuadTree::get_leaf_nodes() {
    if (leaf_) return {this};

    std::vector<QuadTree*> result;
    auto nw = northWest_->get_leaf_nodes();
    auto ne = northEast_->get_leaf_nodes();
    auto sw = southWest_->get_leaf_nodes();
    auto se = southEast_->get_leaf_nodes();
    result.insert(result.end(), nw.begin(), nw.end());
    result.insert(result.end(), ne.begin(), ne.end());
    result.insert(result.end(), sw.begin(), sw.end());
    result.insert(result.end(), se.begin(), se.end());
    return result;
}

size_t QuadTree::get_number_of_agents_in_node() const {
    return particles_.size();
}


} // namespace invertedai
